src/Lean/Elab/Command.lean

```diff
@@ -285,13 +277,38 @@ partial def elabCommand (stx : Syntax) : CommandElabM Unit := do
           | []      => throwError "elaboration function for '{k}' has not been implemented"
           | elabFns => elabCommandUsing s stx elabFns
     | _ => throwError "unexpected command"
+
+/--
+`elabCommand` wrapper that should be used for the initial invocation, not for recursive calls after
+macro expansion etc.
+-/
+def elabCommandTopLevel (stx : Syntax) : CommandElabM Unit := do
+  let initMsgs ← modifyGet fun st => (st.messages, { st with messages := {} })
+  let initInfoTrees ← getResetInfoTrees
+  withLogging do
+    runLinters stx
+  -- We should *not* factor out `elabCommand`'s `withLogging` to here since it would make its error
+  -- recovery more coarse. In particular, If `c` in `set_option ... in $c` fails, the remaining
+  -- `end` command of the `in` macro would be skipped and the option would be leaked to the outside!
+  elabCommand stx
+
+  -- note the order: first process current messages & info trees, then add back old messages & trees,
+  -- then convert new traces to messages
   let mut msgs ← (← get).messages
   -- `stx.hasMissing` should imply `initMsgs.hasErrors`, but the latter should be cheaper to check in general
   if !showPartialSyntaxErrors.get (← getOptions) && initMsgs.hasErrors && stx.hasMissing then
     -- discard elaboration errors, except for a few important and unlikely misleading ones, on parse error
     msgs := ⟨msgs.msgs.filter fun msg =>
       msg.data.hasTag `Elab.synthPlaceholder || msg.data.hasTag `Tactic.unsolvedGoals⟩
-  modify ({ · with messages := initMsgs ++ msgs })
+  for tree in (← getInfoTrees) do
+    let fmt ← tree.format
```

Comment:
Eager formatting?

Reply:
Yes, I forgot that `trace.Elab.info` and `enableInfoTree` are separate... let's wait for #541